core/primitives-core/src/profile.rs

```diff
@@ -180,96 +165,38 @@ pub enum Cost {
-        Cost::ExtCost { ext_cost_kind: ExtCosts::storage_iter_next_base },
-        Cost::ExtCost { ext_cost_kind: ExtCosts::storage_iter_next_key_byte },
-        Cost::ExtCost { ext_cost_kind: ExtCosts::storage_iter_next_value_byte },
-        Cost::ExtCost { ext_cost_kind: ExtCosts::touching_trie_node },
-        Cost::ExtCost { ext_cost_kind: ExtCosts::read_cached_trie_node },
-        Cost::ExtCost { ext_cost_kind: ExtCosts::promise_and_base },
-        Cost::ExtCost { ext_cost_kind: ExtCosts::promise_and_per_promise },
-        Cost::ExtCost { ext_cost_kind: ExtCosts::promise_return },
-        Cost::ExtCost { ext_cost_kind: ExtCosts::validator_stake_base },
-        Cost::ExtCost { ext_cost_kind: ExtCosts::validator_total_stake_base },
-        Cost::WasmInstruction,
-        Cost::ExtCost { ext_cost_kind: ExtCosts::alt_bn128_g1_multiexp_base },
-        Cost::ExtCost { ext_cost_kind: ExtCosts::alt_bn128_g1_multiexp_element },
-        Cost::ExtCost { ext_cost_kind: ExtCosts::alt_bn128_pairing_check_base },
-        Cost::ExtCost { ext_cost_kind: ExtCosts::alt_bn128_pairing_check_element },
-        Cost::ExtCost { ext_cost_kind: ExtCosts::alt_bn128_g1_sum_base },
-        Cost::ExtCost { ext_cost_kind: ExtCosts::alt_bn128_g1_sum_element },
-        #[cfg(feature = "protocol_feature_ed25519_verify")]
-        Cost::ExtCost { ext_cost_kind: ExtCosts::ed25519_verify_base },
-        #[cfg(feature = "protocol_feature_ed25519_verify")]
-        Cost::ExtCost { ext_cost_kind: ExtCosts::ed25519_verify_byte },
-    ];
-
-    pub fn index(self) -> usize {
+    pub fn iter() -> impl Iterator<Item = Self> {
+        let actions =
+            ActionCosts::iter().map(|action_cost_kind| Cost::ActionCost { action_cost_kind });
+        let ext = ExtCosts::iter().map(|ext_cost_kind| Cost::ExtCost { ext_cost_kind });
+        let op = std::iter::once(Cost::WasmInstruction);
+        actions.chain(ext).chain(op)
+    }
+
+    /// Entry in DataArray inside the gas profile.
+    ///
+    /// Entries can be shared by multiple costs.
+    ///
+    /// TODO: Remove after completing all of #8033.
+    pub fn profile_index(self) -> usize {
+        // make sure to increase `DataArray::LEN` when adding a new index
         match self {
             Cost::ActionCost { action_cost_kind: ActionCosts::create_account } => 0,
             Cost::ActionCost { action_cost_kind: ActionCosts::delete_account } => 1,
-            Cost::ActionCost { action_cost_kind: ActionCosts::deploy_contract } => 2,
-            Cost::ActionCost { action_cost_kind: ActionCosts::function_call } => 3,
+            Cost::ActionCost { action_cost_kind: ActionCosts::deploy_contract_base } => 2,
+            Cost::ActionCost { action_cost_kind: ActionCosts::deploy_contract_byte } => 2,
+            Cost::ActionCost { action_cost_kind: ActionCosts::function_call_base } => 3,
+            Cost::ActionCost { action_cost_kind: ActionCosts::function_call_byte } => 3,
```

Comment:
Yeah, I think it's fine to map them to a single entry for the time being. Long term, we should just make the array bigger. The serialization/deserialization code doesn't really care about length, is just zeros unknown cost.

It's better _not_ to change the meaning of indicies, to simplify historical analysis, but even that, strinctly speaking, is not required. If we want, wee can reassign the meaning without protocol change.